sympy/solvers/solveset.py

```diff
@@ -3665,33 +3699,57 @@ def nonlinsolve(system, *symbols):
         return _solveset_work(system, symbols)
 
     # main code of def nonlinsolve() starts from here
-    polys, polys_expr, nonpolys, denominators = _separate_poly_nonpoly(
-        system, symbols)
-
-    if len(symbols) == len(polys):
-        # If all the equations in the system are poly
-        if is_zero_dimensional(polys, symbols):
-            # finite number of soln (Zero dimensional system)
-            try:
-                return _handle_zero_dimensional(polys, symbols, system)
-            except NotImplementedError:
-                # Right now it doesn't fail for any polynomial system of
-                # equation. If `solve_poly_system` fails then `substitution`
-                # method will handle it.
-                result = substitution(
-                    polys_expr, symbols, exclude=denominators)
-                return result
-
-        # positive dimensional system
-        res = _handle_positive_dimensional(polys, symbols, denominators)
-        if res is S.EmptySet and any(not p.domain.is_Exact for p in polys):
-            raise NotImplementedError("Equation not in exact domain. Try converting to rational")
-        else:
-            return res
 
+    polys, polys_expr, nonpolys, denominators, unrad_changed = \
+        _separate_poly_nonpoly(system, symbols)
+
+    poly_eqs = []
+    poly_sol = [{}]
+
+    if polys:
+        # Convert floats to Rational for polynomial calculations
+        polys = [poly(nsimplify(p, rational=True)) for p in polys]
```

Comment:
> Maybe this should be done in _handle_poly and maybe the results should be converted back to Float at the end somehow. Which of the test cases depends on this?

I agree that it would be better to do this in `_handle_poly`.
On master, if the input polynomials have `Float` coefficients and `solve_poly_system` is not able to solve the system, `NotImplementedError` is raised. The following test tests this behavior:
https://github.com/sympy/sympy/blob/56b1c08619dc4c4b53567fca2fffb9ca7ca6445a/sympy/solvers/tests/test_solveset.py#L1587
On this PR, the code that raises `NotImplementedError` is removed and `nsimplify` is used to convert the coefficients to `Rational`

> Perhaps before computing the Groebner basis floats should be checked for. If there are floats then it could convert to rational just while computing the basis and then convert back to float again after

I think this could work. The best way I can think of to convert the polynomials in the Groebner basis back to have `Float` coefficients (this is only needed in the positive-dimensional case, where `_handle_poly` returns the polynomials in the Groebner basis) is to unify the domain with `RR` something like:
```python
>>> p = Poly(x**2 + 2*x + 3*y, x)
>>> Poly(p, domain=p.domain.unify(RR)).as_expr()
1.0*x**2 + 2.0*x + 3.0*y
```
Do you think this is a good idea?

Another approach that might work here is to use polynomials with `Rational` coefficients throughout `nonlinsolve` but to call `evalf` on the final solution if the original polynomials passed to `nonlinsolve` have `Float` coefficients. This will ensure that float solutions are returned if the input polynomials have float coefficients. What do you think about this approach?

Reply:
> The following test tests this behavior:

I don't see the point of that test. I think it can be ignored.

The reason for converting to Rational is that `groebner` can produce incorrect results for float input. I think that the conversion to Rational and back to Float should happen only around the call to Groebner and maybe also solve_poly_system.

For now though for this PR is the float/rational conversion even needed? Which of the tests here would fail if nsimplify was not used?

Reply:
> For now though for this PR is the float/rational conversion even needed? Which of the tests here would fail if nsimplify was not used?

To my knowledge, there are no tests apart from the test I mentioned earlier that would fail if `nsimplify` were not used. Not using `nsimplify` could lead to some incorrect results, though. For example, `nonlinsolve([(x+y)**2 - 9, x**2 - y**2 - 0.75], (x, y))` (the example in the test), would return `EmptySet` even though the system has solutions.

Reply:
It seems that using `groebner` with the `fb5` method handles that case better:
```python
In [15]: groebner([(x+y)**2 - 9, x**2 - y**2 - 0.75], (x, y), method='f5b')
Out[15]: 
             ⎛⎡                                 2           ⎤                           ⎞
GroebnerBasis⎝⎣1.0⋅x - 1.18181818181818⋅y, 1.0⋅y  - 1.890625⎦, x, y, domain=ℝ, order=lex⎠

In [16]: groebner([(x+y)**2 - 9, x**2 - y**2 - 0.75], (x, y))
Out[16]: GroebnerBasis([1.0], x, y, domain=ℝ, order=lex)
```
I wonder if the 'f5b' method generally works better for float input.

In any case I think that if the reason for using `nsimplify` here is just to handle the cases where `groebner` does not work so well with floats then the use of nsimplify should be limited to exactly where `groebner` is called and the output from `groebner` should be converted back to float.

I think that it would be better to handle the use of `nsimplify` in a different PR though. This PR has gone on for long enough and has some good changes that should get merged. The use of `nsimplify` should be considered separately with more than just a single test case for consideration.

I think that the thing to do here is revert the use of `nsimplify` and add a check for this somewhere in `_handle_poly`.